Interrupt subsystem update for cascaded interrupts

lyakh opened the pull request:
The first 35 patches are the actual interrupt subsystem modification to support cascaded interrupt controllers. Then 3 patches re-add the DW UART driver, here just for illustration, since they are handled in #957 . The last 2 patches add a DW INTC driver and enable it on Sue Creek. I expect a couple of comments here and there at least, so I pushed all those patches to this branch. In the final version, of course, they should not be present, at least the UART patches.
This patch series allows the removal of interrupt-map.h, but it is also included from host tools, so I didn't remove it for now. Once the header is removed, we should also remove "select IRQ_MAP" and "config IRQ_MAP" entries from Kconfig files.

RanderWang commented:
@lyakh  It is a huge list. I checked some of them and enjoy them.  Could you split them to small ones ?

singalsu commented:
For what triggered review request for me (the small DMIC related changes) this looks OK. Apparently there's need to do fixes to pass the CI checks.

lyakh commented on Kconfig:
```diff
@@ -58,4 +58,22 @@ config BUILD_VM_ROM
+
+config TRACE_DMA
+	bool "Use DMA to host RAM for tracing"
+	depends on !SUECREEK
```
If you place them in defconfigs, that means, they would be selectable by the user, does that really make sense? What sense can it make to enable TRACE_DMA on a system, that doesn't have that capability? But yes, I can make this a "select" from all platforms apart from SUECREEK, that would be better, I think, thanks.

jajanusz replied:
For capability issue you have `depends on`. Decision to use feature should be selectable.

lyakh commented on src/drivers/Kconfig:
```diff
@@ -96,6 +96,23 @@ endmenu # "Decimation factors"
+	  Enable the DesignWare UART driver
+
+config UART
+	bool
```
The idea is, that the user or "selects" or *_defconfig select what specific hardware is available on the system. Some platforms can have DW_UART, others X_UART, others still Y_UART etc. And there are features in the SOF, that depend on "some" UART. So, all those specific UART implementations select UART to tell the code, that yes, UARTs are available on this system. Why would you make this selectable by the user? Are you proposing to make it possible to select DW_UART and deselect UART? What would that result in?

lyakh commented on src/drivers/dw/intc.c:
```diff
@@ -0,0 +1,159 @@
+{
+	struct irq_cascade_desc *cascade = container_of(desc,
+						struct irq_cascade_desc, desc);
+	trace_point(0x90000 + irq);
```
oops, that's a debug left-over, will remove, thanks for spotting!

lyakh commented:
@RanderWang thanks, I'm glad you liked it :-) But TBH I'm not sure I understand why splitting it would make it easier. Whether you make it 5 series of 7 commits each, or 7 series of 5 commits each or whatever - it's still 35. And they are (almost) all interdependent. So, it wouldn't be possible to, say, extract 10 of them, that are independent, that touch other files. Some of them can be swapped of course, some have no dependencies and can be extracted from the series, but most of it is really a huge gradual change. So, if you split it into 5 series with 7 commits each, you'll have to wait with each series until all previous ones have been reviewed and merged.

tlauda commented on src/drivers/intel/cavs/interrupt.c:
```diff
@@ -73,7 +74,8 @@ static inline void irq_lvl2_handler(void *data, int level, uint32_t ilxsd,
 		list_for_item(clist, &cascade->child[i]) {
```
We iterate over children list, but we do not do cache invalidation before.
You can have out of date data here, if some other core in the meantime registered/unregistered interrupt.

lyakh replied:
On the one hand I'm not changing that, I'm basically just replacing one struct with a different one in this patch. But I think you're right. And it's even worse than that. It isn't just a caching problem, it's a real race: this handler isn't taking the spinlock and the child interrupt registration function isn't using spin_lock_irq(), it's only doing spin_lock(), so, while a process is registering an interrupt, if another handler is already registered for that interrupt, it can trigger, it will interrupt the registration code and can access the corrupted list. Thanks for spotting this!

lyakh replied:
Actually, I think, we didn't need cache-flushing before my patches and also with the first of them, because there interrupt descriptors were per-core. It's only with the second patch in this series we move to a single descriptor for all cores, then we need cache-flushing.

tlauda commented on src/lib/interrupt.c:
```diff
@@ -67,18 +68,22 @@ static int irq_register_child(struct irq_desc *parent, int irq, int unmask,
 	/* increment number of children */
 	cascade->num_children++;
 
+	dcache_writeback_region(cascade, sizeof(*cascade));
```
What with the writebacks after child unregister?

lyakh replied:
Rrright, I think we need that too, thanks!

lgirdwood commented on src/arch/xtensa/smp/idc.c:
```diff
@@ -34,7 +34,7 @@ void idc_enable_interrupts(int target_core, int source_core)
 	idc_write(IPC_IDCCTL, target_core,
 		  IPC_IDCCTL_IDCTBIE(source_core));
-	platform_interrupt_unmask(PLATFORM_IDC_INTERRUPT(target_core));
+	interrupt_cascade_unmask(PLATFORM_IDC_INTERRUPT(target_core));
```
Can we not just call interrupt_unmask(), the clients dont care about the IRQ implementation. ditto for others, we can remove cascade from external API names.

lyakh replied:
Same reasoning basically - I wanted to keep those APIs apart while performing changes. To avoid breaking anything. Once we have a reviewed verified working state, we can merge APIs together.

lgirdwood commented on src/include/sof/math/numbers.h:
```diff
@@ -35,8 +35,16 @@
 
-#define MIN(a, b) (((a) < (b)) ? (a) : (b))
-#define MAX(a, b) (((a) > (b)) ? (a) : (b))
+#define MIN(a, b) ({		\
```
Separate PR for this.

lgirdwood commented on src/drivers/intel/haswell/timer.c:
```diff
@@ -115,10 +115,10 @@ void timer_unregister(struct timer *timer)
 void timer_enable(struct timer *timer)
 {
-	interrupt_enable(timer->irq);
+	interrupt_enable(timer->irq, 0);
```
re-reverts some previous patches.

lyakh replied:
Sorry, I cannot find that in my patches? Maybe an even earlier patch removed that parameter?

lgirdwood commented on src/arch/xtensa/include/arch/timer.h:
```diff
@@ -39,6 +39,7 @@
 struct timer {
 	uint32_t id;
 	uint32_t irq;
+	unsigned int cpu;
```
uint32_t

lyakh replied:
any specific reason for this? I'm pretty sure we aren't going to overflow the smallest imaginable "int" with the number of our cores for a rather long time ;-)

lyakh replied:
uhm, sorry, I disagree... First of all the above void* is not a fixed-width field. Then "irq" to this moment is a bit-mask, so, I can understand why it's fixed-width. ID actually doesn't make sense to be fixed-width to me, but it'll be gone in a couple of patches. Further, I don't think we're going to deal with systems / compilers, on which int isn't 32-bit? But in general, I think the rule in the kernel is, that fixed-width types should *only* be used where required, e.g. when mimicking hardware registers or for similar uses. Other variables should be typed according to their context, meaning - int, long, char, unsigned int etc.

lgirdwood commented on src/audio/host.c:
```diff
@@ -292,7 +292,7 @@ static int host_trigger(struct comp_dev *dev, int cmd)
 		if (ret < 0)
 			trace_host_error("host_trigger(): dma stop failed: %d",
-					 ret)
+					 ret);
```
Separate PR

lgirdwood commented on src/lib/interrupt.c:
```diff
@@ -215,7 +215,7 @@ static uint32_t irq_enable_child(struct irq_desc *parent, int irq,
 	/* enable the parent interrupt */
 	if (parent->enabled_count == 0)
-		arch_interrupt_enable_mask(1 << SOF_IRQ_NUMBER(irq));
+		arch_interrupt_enable_mask(1 << parent->irq);
```
this looks similar to a previous patch and could be merged.

lyakh replied:
You mean patch "interrupt: SOF_IRQ_NUMBER() for DSP interrupts is not needed?" Well, they both get rid of SOF_IRQ_NUMBER(), that's true, but in different ways and for different reasons, so, I think it's clearer if they're kept apart.

lgirdwood commented on src/lib/schedule.c:
```diff
@@ -51,6 +51,7 @@ struct schedule_data {
 	struct list_item list;	/* list of tasks in priority queue */
 	uint32_t clock;
 	struct work work;
+	int irq;
```
net loss of 4 bytes for no gain ?

lyakh replied:
This is needed, because interrupts will become virtual later. They will be mapped at run-time based on their interrupt controller parent and the physical interrupt number on that controller. So, we want to map interrupts once and then use the virtual IRQ number. See patch https://github.com/thesofproject/sof/pull/1053/commits/493b086682d398b95fb76ab5095f8049c57a0d43

lgirdwood replied:
Ok, but is the virtual number constant for each platform. i.e. APL virt IRQ 10 is always SSP0 ? If not then this will be painful to debug.

lyakh replied:
As long as the board configuration doesn't change, yes, they will stay constant. The only thing, that can change them is adding new or removing existing interrupt controllers. I don't think that would happen too often. We can add trace to the mapping function, as long as the IPC interrupt is a DSP native one. Or we can dump a mapping at the end of booting and then print any new mappings.

lgirdwood commented on src/include/sof/idc.h:
```diff
@@ -115,6 +115,7 @@ struct idc {
 	uint32_t done_bit_mask;		/**< done interrupt mask */
 	struct idc_msg received_msg;	/**< received message */
 	struct task idc_task;		/**< IDC processing task */
+	int irq;
```
ditto, if compiler is providing this then there is no need to store it locally.

lgirdwood commented on src/drivers/intel/haswell/timer.c:
```diff
@@ -115,10 +115,10 @@ void timer_unregister(struct timer *timer)
 void timer_enable(struct timer *timer)
 {
-	interrupt_enable(timer->irq, 0);
+	interrupt_enable(timer->irq, NULL, 0);
```
Lets not use NULL since its being used internally as a UUID.

lgirdwood commented on src/arch/xtensa/task.c:
```diff
@@ -176,6 +176,7 @@ int arch_run_task(struct task *task)
 
 int arch_allocate_tasks(void)
 {
+	unsigned int cpu = cpu_get_id();
```
The CPU core for each task will only be known at task creation time. i.e. after topology has been loaded. So we cannot hard code to current CPU unless this is being run on ALL CPUs.

lyakh replied:
@lgirdwood Sorry, where is this done? I see
```
        interrupt_register(PLATFORM_IRQ_TASK_LOW, IRQ_AUTO_UNMASK, _irq_task,
                           task_irq_low_get());
```
where PLATFORM_IRQ_TASK_LOW is IRQ_NUM_SOFTWARE[234] on all platforms, i.e. a DSP interrupt. Similar for MED and HIGH. That means they are registered on the current CPU. What am I missing?

lgirdwood replied:
Not yet implemented on other cores atm, topology can request core > 0 but code will assign 0 (although this will change with upcoming PRs).

lyakh replied:
hm, which PR is this? I think it should actually become better with these changes: instead of mangling on bitfields you'd call interrupt_enable() with a specific CPU, that's it. And it would be a good test case for that!

lgirdwood replied:
PRs for scheduling pipelines on other cores are in progress by @tlauda @mmaka1 , I dont know when they will be ready for merging though.

lyakh replied:
would it be possible to merge this first and then port their work on top?

lyakh commented on src/lib/interrupt.c:
```diff
@@ -33,11 +33,76 @@
+
+	spin_lock_irq(&cascade_lock, flags);
+
+	list_for_item (list, &cascade_list) {
```
Oh... So, basically, any memory, that can be accessed from multiple cores, has to be cache-invalidated before reading and cache-flushed after writing?.. And when scanning lists, that means cache-invalidating each list item before reading?.. If yes, I think we need SMP-safe list accessors...

lgirdwood replied:
yep, we could have a spin_lock_irq_invalidate(lock, memory_to_invalidate, size);

lgirdwood replied:
and also a list_for_item_invalidate()

lyakh commented on src/drivers/intel/baytrail/interrupt.c:
```diff
@@ -102,7 +102,7 @@ uint32_t platform_interrupt_get_enabled(void)
 }
 
-void platform_interrupt_mask(uint32_t irq)
+void baytrail_interrupt_mask(uint32_t irq)
```
I thought it could be controversial. TBH, by this I made my life a bit easier. I moved byt and hsw out of the way "temporarily" to safely concentrate on the only cascading arch - cAVS. This way we're sure not to break hsw and byt on this long journey. So, would it be acceptable to do it this way, but in the end add 2 more patches to return byt and hsw back to the generic APIs? Would keep the series a bit easier really...

lgirdwood replied:
We need to be consistent across all driver using the same APIs, this would just add extra unessesary APIs to learn and document.

abonislawski commented on src/arch/xtensa/boot_loader.c:
```diff
@@ -276,11 +277,25 @@ static int32_t lp_sram_init(void)
 }
 #endif
 
+#if defined(CONFIG_TRACE_UART)
```
Please start using #if (CONFIG_*) instead of _defined_, like in recent changes in sof (#if defined CONFIG_ *-> #if CONFIG_*)

lyakh commented on src/drivers/intel/cavs/dmic.c:
```diff
@@ -1505,9 +1505,6 @@ static void dmic_irq_handler(void *data)
 				 "full fifo B or PDM overrrun");
-
-	/* clear IRQ */
-	platform_interrupt_clear(dmic_irq(dai), 1);
```
This is a cAVS-specific driver and on cAVS platform_interrupt_clear() was a dummy.

lyakh commented on src/drivers/intel/baytrail/ssp.c:
```diff
@@ -622,7 +622,7 @@ static int ssp_probe(struct dai *dai)
 	}
 
-	platform_interrupt_unmask(ssp_irq(dai), 1);
+	platform_interrupt_unmask(ssp_irq(dai));
```
I don't think so, this is just patch 08/40, it explicitly removes this parameter, it wasn't added before. Or have I misunderstood something?

lyakh commented on src/drivers/intel/haswell/ipc.c:
```diff
@@ -99,15 +99,15 @@ static void irq_handler(void *arg)
 		/* Mask Done interrupt before return */
 		shim_write(SHIM_IMRD, shim_read(SHIM_IMRD) | SHIM_IMRD_DONE);
-		interrupt_clear(PLATFORM_IPC_INTERRUPT);
+		interrupt_clear(PLATFORM_IPC_INTERRUPT, 1);
```
It is passed on to platform_interrupt_clear(), and there it is used on byt.

lyakh commented:
@lgirdwood what's your preferred way to proceed with this PR? I realise that it is rather large, but splitting it doesn't really make it easier IMHO. Bear in mind, that the last 5 patches don't really belong to the PR. As for squashing - I think that would make patches less clear. I can see a couple of patches that are already a bit bigger than what one would normally want to have, but I don't see any patches, that are too small :-) Github really isn't the best tool for such reviews... I can email them, if that helps?

lgirdwood commented:
> @lgirdwood what's your preferred way to proceed with this PR? I realise that it is rather large, but splitting it doesn't really make it easier IMHO. Bear in mind, that the last 5 patches don't really belong to the PR. As for squashing - I think that would make patches less clear. I can see a couple of patches that are already a bit bigger than what one would normally want to have, but I don't see any patches, that are too small :-) Github really isn't the best tool for such reviews... I can email them, if that helps?

First step is to address comments and resend. We don't need the macro updates/fixes, they can go in another PR that can be merged immediately. Dont need the UART driver either in the PR (it can be a another PR).

lyakh commented:
@lgirdwood sure, most comments are non-controversial and easy to address. I'll take out two macro-fixing patches as well of course. I see at least two larger issues:
1. cache. I've implemented smp-safe list-manipulation primitives like "for_each" or "add" or "delete." But all that has to be carefully double-checked...
2. have I managed to convince you to let me *temporarily* use platform-specific functions for interrupt (un)masking, enabling / disabling etc? I think it really makes the series easier to make and verify. That way we know, that we don't break those systems. After we've converted interrupt numbers, we should bring those drivers back to using standard methods.

lgirdwood commented:
> 2\. have I managed to convince you to let me _temporarily_ use platform-specific functions for interrupt (un)masking, enabling / disabling etc? I think it really makes the series easier to make and verify. That way we know, that we don't break those systems. After we've converted interrupt numbers, we should bring those drivers back to using standard methods.

No, changing an API to later change it back makes no sense. It only creates more uncertainty.

lyakh commented:
> > 2. have I managed to convince you to let me _temporarily_ use platform-specific functions for interrupt (un)masking, enabling / disabling etc? I think it really makes the series easier to make and verify. That way we know, that we don't break those systems. After we've converted interrupt numbers, we should bring those drivers back to using standard methods.
> 
> No, changing an API to later change it back makes no sense. It only creates more uncertainty.

It isn't very pretty, I agree, but I thought easier development, debugging and review were a good reason for that. But np, I can redo that part.

lyakh commented:
@tlauda I'm beginning (too late but still) to doubt: are we really sure that caches aren't coherent on xtensa between cores?

tlauda commented:
I'm sure.

lyakh commented:
@tlauda that's a pity... Do we have any datasheets where this is described?

lyakh commented:
@tlauda @lgirdwood this manual cache management really gets very intrusive, error-prone, pollutes the sources a lot, and probably causes non-negligible performance reduction. Are there other ways to handle that? If we collect all that data, that we know, that different CPUs will access in a common area or even section, maybe we can simplify that management. Is it possible to make that area non-cacheable? That would possibly cost even more performance-wise. How are caches invalidated / written back - with a cache-line granularity? We could try to manually optimise for cache-line lengths, but as any manual optimisation, that might be hard to do right and to keep maintaining. Or we could manually flush / invalidate that whole area when beginning to handle any of our SMP-critical data and before returning from those functions. Or would flushing a couple of K of data be worse than repeated flushing of a handful of bytes? Any other options?

lgirdwood commented:
@lyakh manual cache management is unavoidable. It compiles out on platforms that manage it in HW.  Having cache aware list accessors and locks APIs will simplify things a lot, so these should probably be created first as a new PR.

e.g.
```
spin_lock_invalidate(lock, object, object_size);

list_for_item_invalidate(item_ptr, list, item_size)
```

We also need to be aware that some operation are global (like IRQ registration) and some are local like the IRQ handler flow and this would influence what would need cache ops i.e. IRQ handler would not need to invalidate global IRQ list but rather a sub list/object for that IRQ/level.

lyakh commented:
This is an update with cleaned up cache handling, now using added SMP-safe list manipulation primitives and also synchronising the data cache manually as required. This now doesn't include the DW UART and the DW INTC patches, so, in principle this is the set, that has to be merged.

lgirdwood commented on src/drivers/intel/cavs/interrupt.c:
```diff
@@ -213,13 +213,13 @@ uint32_t platform_interrupt_get_enabled(void)
 void platform_interrupt_set(uint32_t irq)
 {
-	if (!interrupt_get_parent(irq))
+	if (interrupt_is_dsp(irq))
```
Lets make this more meaningful, interrupt is_dsp_direct() implies it's a direct (no external IP) IRQ type.

lgirdwood commented on src/include/sof/interrupt.h:
```diff
@@ -114,6 +114,8 @@ int interrupt_cascade_register(struct irq_cascade_desc *cascade);
 struct irq_cascade_desc *interrupt_get_parent(uint32_t irq);
 int interrupt_get_irq(unsigned int irq, const char *cascade);
 
+#define interrupt_is_dsp(irq) (irq < PLATFORM_IRQ_CHILDREN)
```
A comment is needed here to explain the macro.

lgirdwood commented:
@lyakh @mengdonglin this needs QA validation on APL, BYT, BDW prior to merge after review updates and build fixes are pushed.

lyakh commented on src/drivers/intel/cavs/interrupt.c:
```diff
@@ -69,11 +70,15 @@ static inline void irq_lvl2_handler(void *data, int level, uint32_t ilxsd,
 		if ((status & 0x1) == 0)
 			goto next;
 
+		spin_lock(&cascade->lock);
```
it's per cascade-controller. It's only protesting child lists of that controller, so, I don't think they can have different priorities?

lgirdwood replied:
ok, as long as we cant change these list from higher level IRQs. This assumption should be stated in comments so it's obvious.

lyakh commented on src/include/sof/interrupt.h:
```diff
@@ -59,12 +59,16 @@ struct irq_desc {
 	struct list_item irq_list;
+};
 
+struct irq_cascade_desc {
```
Sorry, with this gh commenting I cannot figure out to which parts your above comments refer? The first one is for the entire struct, but the second one?

lgirdwood replied:
entire struct, but I see you have comments now.

lyakh commented on src/include/sof/list.h:
```diff
@@ -72,6 +74,18 @@ static inline void list_item_append(struct list_item *item,
+{
+	list_item_append(item, list);
+
+	dcache_writeback_region(list, sizeof(*list));
```
I was considering list_item_append_invalidate_writeback(), but in the end I decided, that I only had 1 or maybe a couple of uses for that, and that it's actually dangerous, it can drop your recent changes. Let's not add it in this patch, later if it turns out we have sufficiently many uses for it, we can add it.

lgirdwood replied:
ok for list_item_append_invalidate_writeback() to be dropped.

lyakh commented on src/lib/interrupt.c:
```diff
@@ -87,34 +95,40 @@ static int irq_register_child(struct irq_desc *parent, int irq, int unmask,
 						struct irq_cascade_desc, desc);
+	struct list_item *clist = &cascade->child[SOF_IRQ_BIT(irq)];
 
 	spin_lock(&cascade->lock);
```
This isn't changed by this commit and not in this series at all. this spin-lock protects lists of IRQ children, I don't think those lists are ever modified from an IRQ context? If they are, we certainly should fix this, but this can be a separate patch.

lgirdwood replied:
best to rebase the fix into this one otherwise we can break bisect (and make it confusing for any debugger/tester).

lyakh commented on src/lib/interrupt.c:
```diff
@@ -33,12 +33,89 @@
+static spinlock_t cascade_lock;
+static struct list_item cascade_list = {&cascade_list, &cascade_list};
+
+int interrupt_cascade_register(struct irq_cascade_desc *cascade)
```
No! This is *the* function, that is used to register cascaded interrupt controllers.

lyakh commented on src/lib/interrupt.c:
```diff
@@ -44,16 +44,25 @@ static int irq_register_child(struct irq_desc *parent, int irq, int unmask,
 
-	if (parent == NULL)
-		return -EINVAL;
+	spin_lock(&cascade->lock);
```
yes, not changed here, let me add a patch at the end...

lyakh commented on src/include/sof/dmic.h:
```diff
@@ -333,6 +333,7 @@ struct dmic_pdata {
 	struct work dmicwork;
 	int32_t startcount;
 	int32_t gain;
+	unsigned int cpu;
```
Let's discuss this separately...

lyakh commented on src/include/sof/interrupt.h:
```diff
@@ -111,7 +111,7 @@ uint32_t interrupt_disable(uint32_t irq, void *arg, unsigned int cpu);
 void interrupt_init(void);
 int interrupt_cascade_register(struct irq_cascade_desc *cascade);
-struct irq_desc *interrupt_get_parent(uint32_t irq);
+struct irq_cascade_desc *interrupt_get_parent(uint32_t irq);
```
irq_desc is for interrupt "consumers," for single interrupts, irq_cascade_desc is for cascading IRQ controllers

lgirdwood replied:
ok, it would be nice to split the headers here for consumers and controllers (but that could come as a new PR).

lgirdwood commented:
@lyakh please let @mengdonglin know when this is ready for QA,(with no conflicts)  we can still do this while reviewing is in progress (and we are almost there now). @mengdonglin this will need tested on platforms, it's a big update but very needed.

lyakh commented:
@lgirdwood I've just rebased, re-tested and re-pushed the stack. All patches applied cleanly. @mengdonglin please take it for a spin ASAP, before a random unrelated commit creates a conflict.

tlauda commented:
@lgirdwood @mengdonglin Please also verify multicore functionality. This set of patches touch interrupt handling for multiple cores very widely. We could have missed cache handling somewhere and caused the regression.

keqiaozhang commented:
I tested this PR on GLK and found 5 regressions:
1. HDMI playback failed with I/O error.
2. DMIC record failure, no data recorded.
3. Volume control doesn’t work via GPA1.*
    How to reproduce:
     a. Use aplay to play a wav file.
     b. Open another terminal and change the corresponding PGA * volume kcontrol in alsamixer, but 
          volume doesn’t change accordingly.
4. Playback failed to continue form backend.
    How to reproduce:
    a. Use aplay to play a wav file, then press ctl+x to stop aplay.
    b. Press “fg” to resume aplay, after this, no sound can be heard.
5. Sof-logger: no trace log output 
    How to reproduce:
    a. run sof-logger: sof-logger -l sof-apl.ldc –t

Test recipe:
Sof/tplg: sof-master/ 052942 + PR1053
Kernel driver: cros-sof-v4.14-rebase/ a1d6bf0

stevyan commented:
Following is the PR1053 on APL UP2 pcm512x, 3 regressions found as blow:
•	HDMI playback failed with “Unable to install hw params (-22)” error;
•	Volume control doesn’t work via <PGA1.0 1>, but the ‘Digital’ can control the volume;
•	sof-logger: no trace log output 

**Test Env:**
sof/tplg: sof-master(052942c8) + PR1053
Kernel driver: topic/sof-dev (f304b5c2)

stevyan commented:
Following is the PR1053 on BTY Minnowboard rt5651, 1 regressions found as blow:
•	sof-logger: no trace log output 

**Test Env:**
sof/tplg: sof-master(052942c8) + PR1053
Kernel driver: topic/sof-dev (f304b5c2)

lgirdwood commented:
@stevyan can I confirm you are testing APL in SMP mode ?

lgirdwood commented:
@stevyan any update on your testing regarding SMP mode ?

tlauda commented:
@stevyan By testing SMP we are talking about trying to run pipeline on the slave cores, not only testing FW binary build with SMP arch setting.

stevyan commented:
@lgirdwood  @tlauda  I'll schedule the APL SMP mode test for this PR today, and update the corresponding status after testing done.

Jiangxinx commented:
@lgirdwood @tlauda @stevyan This PR can not be added manually,because of the conflicting files.So i can't test the base function on APL.
***console output***
git am 1053.patch
```Applying: list: add SMP-safe list primitives
Applying: interrupt: introduce a separate struct for cascading interrupts
Applying: cavs: eliminate per-core interrupt controller descriptors
Applying: interrupt: add a parameter to interrupt_unregister()
error: patch failed: src/drivers/dw/dma.c:1415
error: src/drivers/dw/dma.c: patch does not apply
Patch failed at 0004 interrupt: add a parameter to interrupt_unregister()
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

lyakh commented:
It has been decided to pursue this approach of common for all DSP cores IRQ controller descriptors. The difference to the first version is, that we aren't trying to maintain cache coherence for all the descriptors manually, which has proven to be problematic, instead we allocate all the descriptors from non-cacheable RAM.

lyakh commented:
@lgirdwood in a comment to #1355 you asked for a high-level comparison. Something like this?

1. Supported features:

- current:

Two levels of interrupt handling are supported on cAVS platforms. The first level is implemented in the XTOS core, the second level is implemented in the SOF. The implementation uses bitmasks as interrupt IDs. Those bitmasks encode the first-level interrupt number of the parent interrupt, the second level interrupt number, the DSP core number, on which this interrupt should be run, a unique ID to distinguish between shared interrupts and the second level IRQ controller ID.

The main elements of the implementation are a parent-lookup function and an interrupt handler.

- new:

Arbitrary IRQ controller levels are supported. IRQ controllers are registered dynamically. Each controller receives a base IRQ number and can handle up to PLATFORM_IRQ_CHILDREN child interrupts. Virtual interrupt numbers are simply a sum of the controller base IRQ number and an IRQ number on that controller. The native DSP interrupts get numbers 0..PLATFORM_IRQ_CHILDREN - 1, interrupts on the first cascading controller numbers PLATFORM_IRQ_CHILDREN..2 * PLATFORM_IRQ_CHILDREN - 1, etc.

The parent-lookup function is now generic, it scans all the registered controllers and checks their virtual interrupt ranges. Additionally interrupt controllers now contain operations structs, currently consisting of a .mask() and an .unmask() function.

2. Initialisation:

- current:

No initialisation is required, the system is hard-coded to use the parent lookup function.

- new:

During initialisation all the cascading interrupt controllers are registered, but in principle this can also be done any time later. At the moment unregistering cascading controllers isn't supported.

3. Memory allocation:

- current:

The interrupt management subsystem uses an IRQ descriptor for all interrupts, includung parent interrupts of cascading interrupt controllers. That descriptor contains an array of list heads for lists of child interrupt descriptors. That means, that even simple device interrupts, unable to host child interrupts, also include those arrays. cAVS cascading interrupt controllers have a capability to
enable and disable child interrupts per DSP core. Using this capability the current implementation contains cascading interrupt controller descriptors for each core.

- new:

We cannot rely on all generic cascading interrupt controllers to be able to manage child interrupts on a per-core basis. This makes the current approach impossible. A new descriptor struct is introduced for cascading controllers to avoid wasting memory as described above.

4. List management:

- current:

Cascading interrupt controllers are statically allocated in cAVS-specific code in a 2-dimensional array over interrupts and DSP cores.

- new:

A global single-linked list of all cascading controllers is used. This is one of the few objects, whose cache has to be synchronised manually.

5. Cache management:

- current:

Descriptors are per-core, this avoids the need to manually synchronise their CPU cache.

- new:

To avoid manual cache synchronisation all interrupt descriptors are now allocated from non-cacheable memory. A small static object is created, containing a head of the descriptor list and a current virtual interrupt base. Cache of this object has to be managed manually.

lyakh commented:
Build failure for "host":
```
/home/sof/work/sof.git/src/include/sof/interrupt.h:35:19: error: 'CONFIG_CORE_COUNT' undeclared here (not in a function); did you mean 'CONFIG_COMP_TONE'?
  int enable_count[CONFIG_CORE_COUNT];
```
Shouldn't CONFIG_CORE_COUNT be defined for host builds too?

jajanusz commented:
> Build failure for "host":
> 
> ```
> /home/sof/work/sof.git/src/include/sof/interrupt.h:35:19: error: 'CONFIG_CORE_COUNT' undeclared here (not in a function); did you mean 'CONFIG_COMP_TONE'?
>   int enable_count[CONFIG_CORE_COUNT];
> ```
> 
> Shouldn't CONFIG_CORE_COUNT be defined for host builds too?
> **Edit:** looks like it shouldn't. There are also PLATFORM_CORE_COUNT and MAX_CORE_COUNT, none of them is defined on "host," MAX_... also isn't defined on imx8.

`CORE_COUNT` is only defined for xtensa, we don't care about cores on host, because we may even not be able to control what we run there, better to stick to OS threads or smth like that.

First of all you should consider if there is no better place for things that depend on that config (outside of project-wide sof includes). If there is really no better place, then please guard it with some meaningful config.

lyakh commented:
> First of all you should consider if there is no better place for things that depend on that config (outside of project-wide sof includes). If there is really no better place, then please guard it with some meaningful config.

@jajanusz until now none of the code, using *_CORE_COUNT macros was built for host. The last patch in this series changes that. It adds such a macro to src/include/sof/interrupt.h which is also included in host builds. I could imagine adding something like
```
#ifndef PLATFORM_CORE_COUNT
#define PLATFORM_CORE_COUNT 1
#endif
```
somewhere, or would that be unacceptable? I wouldn't want to enclose the actual struct fields or whole structs into similar "#ifdef's"

jajanusz commented:
> > First of all you should consider if there is no better place for things that depend on that config (outside of project-wide sof includes). If there is really no better place, then please guard it with some meaningful config.
> 
> @jajanusz until now none of the code, using *_CORE_COUNT macros was built for host. The last patch in this series changes that. It adds such a macro to src/include/sof/interrupt.h which is also included in host builds. I could imagine adding something like
> 
> ```
> #ifndef PLATFORM_CORE_COUNT
> #define PLATFORM_CORE_COUNT 1
> #endif
> ```
> 
> somewhere, or would that be unacceptable? I wouldn't want to enclose the actual struct fields or whole structs into similar "#ifdef's"

Problem with that is that it's lie, because it will probably have more cores. However we have already similar situation with UP builds, imho it's kinda misnamed.

I suggest you to just add `#define PLATFORM_CORE_COUNT 1` to `libary/include/platform/platform.h`, just like it's done for baytrail/haswell. And ten use PLATFORM_CORE_COUNT instead of CONFIG_CORE_COUNT in *global* sof headers.
It's not ideal, but at least is not creating additional mess.

lyakh commented on src/drivers/intel/baytrail/interrupt.c:
```diff
@@ -25,6 +25,14 @@ void platform_interrupt_set(int irq)
 	switch (irq) {
+	case IRQ_NUM_EXT_PMC:
+	case IRQ_NUM_EXT_IA:
+	case IRQ_NUM_SOFTWARE2:
```
@lgirdwood Sorry, on  baytrail and haswell I only moved this code around without changing functionality, I think this was the same before these patches.

lyakh commented on src/drivers/intel/haswell/interrupt.c:
```diff
@@ -30,6 +30,9 @@ void platform_interrupt_clear(uint32_t irq, uint32_t mask)
 	case IRQ_NUM_EXT_SSP0:
 	case IRQ_NUM_EXT_SSP1:
+	case IRQ_NUM_EXT_IA:
+	case IRQ_NUM_SOFTWARE1:
```
Same on haswell

jajanusz replied:
@lyakh Can you add `/* FALLTHRU */` to conform to -Wimplicit-fallthrough as f.e. our tools do.

lyakh commented on src/include/sof/interrupt.h:
```diff
@@ -45,15 +45,27 @@ struct irq_desc {
 
+/* A descriptor for cascading interrupt controller template */
+struct irq_cascade_tmpl {
+	const char *name;
```
Name is how I identify cascading interrupt controllers. To get a virtual IRQ number you provide a name of the controller and its local hardware IRQ number

lgirdwood replied:
Integer ID is easier/faster to check than a name. It can also be used directly by trace and as an array index too if needed.

lyakh commented on src/lib/interrupt.c:
```diff
@@ -8,11 +8,96 @@
+
+	dcache_invalidate_region(&cascade_root, sizeof(cascade_root));
+
+	for (cascade = cascade_root.list; cascade; cascade = cascade->next)
```
@lgirdwood why? It only has one operator - the "if." I really don't think patterns like
```
if () {
	for () {
		if () {
			op1();
			op2();
		}
	}
}
```
look very pretty or helpful. And I haven't seen such requirements in any of the "style" documents nor have I seen checkpatch complain about them...

lgirdwood replied:
@lyakh so I can read it easier and it keeps the {} indentation correct without gaps.

ClarexZhou commented:
Test is done on Firmware:  https://github.com/lyakh/sof/tree/interrupt-whl 2315c28 on all below platforms.  No new issues, compare with Master branch.

BTY   MinnowBoard | BTY   MinnowBoard | APL   UP2 | APL   UP2 | GLK | CML_Mantis | CFL_Nebula | ICL_RVP | WHL_RVP
-- | -- | -- | -- | -- | -- | -- | -- | --
nocodec | I2S   rt5651 | nocodec | I2S   PCM512x | I2S-DA7219 | HDA   ALC3204 | HDA   ALC3254 | HDA   ALC700 | I2S   ALC5682

lgirdwood commented:
@ClarexZhou thanks for the test. @lyakh fyi test results are good.

tlauda commented:
@lgirdwood @lyakh I would just like to confirm it doesn't break SMP.
@zrombel Could you try this with multicore tests?

zrombel commented:
As CI non-blocking tests shows miltikernel tests fails on both ICL and WHL platforms. I run some test manually on those platforms and tests fails after sending trigger start - FW responds with -62 error code.
I can share tests logs and trace logs.

tlauda commented:
@lyakh I'm confused right now. Which PR is the final one? This one or #1355?

lyakh commented:
Having passed all tests (in the sense, of no regressions), I've rebased this series on top of the current sof-dev, For what I can tell, it's now ready to be merged @lgirdwood @tlauda

tlauda commented:
@lyakh This PR still fails multicore tests. Trigger stream IPC throws error -62, when pipeline is configured on slave core.

lyakh commented:
> @lyakh This PR still fails multicore tests. Trigger stream IPC throws error -62, when pipeline is configured on slave core.

@tlauda is that one of the CI tests that show as failed? Do you understand what are those failing tests - I looked at details and don't see any failures. As for your test - I need details - logs and how to reproduce. Without these patches the test is successful I understand?

lyakh commented:
@tlauda I just tried a primitive simultaneous aplay -> SSP + speaker-test -> HDMI on an SMP build on Up2 - it worked. Need instructions to reproduce your failure.

zrombel commented:
@lyakh : on CI multikernel test run on ICL and WHL platforms, and as @tlauda mentioned we get error -62 on stream trigger_start. IPC logs and trace logs attached below:
[dma_trace_1.log](https://github.com/thesofproject/sof/files/3381950/dma_trace_1.log)
[ipc.log](https://github.com/thesofproject/sof/files/3381951/ipc.log)

I will run tests on APL platform and get back with feedback.

zrombel commented:
I build firmware with SMP for Apollolake and I have same result as on ICL and WHL.
With firmware build from master branch multikernel tests pass on all platforms.

keyonjie commented:
@zrombel @lyakh please align with the topology you are using, make sure we have used non-core0 for at least one of the pipeline.

lyakh commented:
I've fixed problems with SMP configurations - thanks again to @tlauda for testing and debugging! To make review easier I'm attaching a diff that collects all the fixes in this version compared to the previous version.
[20190719.diff.txt](https://github.com/thesofproject/sof/files/3411505/20190719.diff.txt)

tlauda commented on src/drivers/interrupt.c:
```diff
@@ -241,14 +291,15 @@ static uint32_t irq_disable_child(struct irq_desc *parent, int irq, void *arg)
 	spin_lock(&cascade->lock);
 
-	child = cascade->child + SOF_IRQ_BIT(irq);
+	child = cascade->child + hw_irq;
```
Remove #include <sof/drivers/interrupt-map.h>

tlauda commented on src/include/sof/drivers/interrupt.h:
```diff
@@ -118,11 +120,13 @@ void interrupt_unmask(uint32_t irq, unsigned int cpu);
  * have SOF_IRQ_LEVEL(irq) != 0.
  */
-#define interrupt_is_dsp_direct(irq) (!SOF_IRQ_LEVEL(irq))
+#define interrupt_is_dsp_direct(irq) (!PLATFORM_IRQ_CHILDREN || \
```
Change comment and also remove #include <sof/drivers/interrupt-map.h>

tlauda commented on src/include/sof/lib/dma.h:
```diff
@@ -223,8 +224,10 @@ void dma_put(struct dma *dma);
 	dma->plat_data.base
-#define dma_irq(dma, cpu) \
-	(dma->plat_data.irq + (cpu << SOF_IRQ_CPU_SHIFT))
+#define dma_irq(dma) \
```
Remove #include <sof/drivers/interrupt-map.h>

tlauda commented on src/include/sof/schedule/ll_schedule.h:
```diff
@@ -36,6 +36,4 @@ struct ll_task_pdata {
 	uint32_t flags;
 };
 
-extern struct timesource_data platform_generic_queue[];
```
Remove #include <sof/drivers/timer.h>.

tlauda commented on src/platform/apollolake/include/platform/lib/dma.h:
```diff
@@ -49,8 +49,7 @@
 
-#define dma_chan_irq(dma, cpu, chan) \
-	(dma_irq(dma, cpu) + (channel << SOF_IRQ_BIT_SHIFT))
+#define dma_chan_irq(dma, chan) (dma_irq(dma) + channel)
```
Remove #include <sof/drivers/interrupt-map.h>.

tlauda commented on src/platform/apollolake/include/platform/platform.h:
```diff
@@ -39,17 +39,23 @@ struct timer;
 /* pipeline IRQ */
-#define PLATFORM_SCHEDULE_IRQ	IRQ_NUM_SOFTWARE4
+#define PLATFORM_SCHEDULE_IRQ		IRQ_NUM_SOFTWARE4
+#define PLATFORM_SCHEDULE_IRQ_NAME	NULL
```
#include <stddef.h>

tlauda commented on src/platform/baytrail/include/platform/platform.h:
```diff
@@ -44,17 +44,22 @@ struct timer;
 
 /* IPC Interrupt */
 #define PLATFORM_IPC_INTERRUPT	IRQ_NUM_EXT_IA
+#define PLATFORM_IPC_INTERRUPT_NAME	NULL
```
#include <stddef.h>

tlauda commented on src/platform/cannonlake/include/platform/platform.h:
```diff
@@ -49,17 +49,23 @@ struct timer;
 
 /* pipeline IRQ */
 #define PLATFORM_SCHEDULE_IRQ	IRQ_NUM_SOFTWARE4
+#define PLATFORM_SCHEDULE_IRQ_NAME	NULL
```
#include <stddef.h>

tlauda commented on src/platform/haswell/include/platform/platform.h:
```diff
@@ -40,17 +40,22 @@ struct timer;
 
 /* IPC Interrupt */
 #define PLATFORM_IPC_INTERRUPT	IRQ_NUM_EXT_IA
+#define PLATFORM_IPC_INTERRUPT_NAME	NULL
```
#include <stddef.h>

tlauda commented on src/platform/icelake/include/platform/platform.h:
```diff
@@ -49,17 +49,23 @@ struct timer;
 
 /* pipeline IRQ */
 #define PLATFORM_SCHEDULE_IRQ	IRQ_NUM_SOFTWARE4
+#define PLATFORM_SCHEDULE_IRQ_NAME	NULL
```
#include <stddef.h>

tlauda commented on src/platform/imx8/include/platform/platform.h:
```diff
@@ -25,18 +25,23 @@ struct timer;
 /* IPC Interrupt */
-#define PLATFORM_IPC_INTERRUPT	IRQ_NUM_MU
+#define PLATFORM_IPC_INTERRUPT		IRQ_NUM_MU
+#define PLATFORM_IPC_INTERRUPT_NAME	NULL
```
#include <stddef.h>

tlauda commented on src/platform/suecreek/include/platform/platform.h:
```diff
@@ -49,17 +49,19 @@ struct timer;
 /* pipeline IRQ */
-#define PLATFORM_SCHEDULE_IRQ	IRQ_NUM_SOFTWARE4
+#define PLATFORM_SCHEDULE_IRQ		IRQ_NUM_SOFTWARE4
+#define PLATFORM_SCHEDULE_IRQ_NAME	NULL
```
#include <stddef.h>

tlauda commented on src/schedule/edf_schedule.c:
```diff
@@ -6,7 +6,6 @@
 
 #include <sof/common.h>
 #include <sof/drivers/interrupt.h>
-#include <sof/drivers/interrupt-map.h>
```
It's still used in this file: SOF_IRQ_PASSIVE_LEVEL.

tlauda commented on src/include/sof/drivers/interrupt.h:
```diff
@@ -36,18 +36,23 @@ struct irq_child {
 	struct list_item list;
 };
 
+/**
```
Doxygen standard goes like this:
```
/** \brief Interrupt client descriptor. */
struct irq_desc {
        int irq;                                   /**< virtual IRQ number */
        void (*handler)(void *arg);   /**< interrupt handler function */
        ...
};
```

lyakh replied:
@tlauda indeed, the style I used is from the Linux kernel, it isn't the classic doxygen style. It looks like the "pure" doxygen doesn't define a way to document struct / class members in a header, not inline. So, we'd end up with
```
/**
 * \brief cascading interrupt controller descriptor.
 */
struct irq_cascade_desc {
	const char *name;				/**< name of the
							 * controller
							 */
	const struct irq_cascade_ops *ops;		/**< cascading interrupt
							 * controller driver
							 * operations
							 */
	struct irq_desc desc;				/**< the interrupt, that
							 * this controller is
							 * generating
							 */
	struct irq_cascade_desc *next;			/**< link to the global
							 * list of interrupt
							 * controllers
							 */
	bool global_mask;				/**< the controller
							 * cannot mask input
							 * interrupts per core
							 */
	spinlock_t lock;				/**< protect child
							 * lists, enable and
							 * child counters
							 */
	int enable_count[PLATFORM_CORE_COUNT];		/**< enabled child
							 * interrupt counter
							 */
	unsigned int num_children[PLATFORM_CORE_COUNT];	/**< number of children
							 */
	struct irq_child child[PLATFORM_IRQ_CHILDREN];	/**< array of child
							 * lists - one per
							 * multiplexed IRQ
							 */
};
```
Is this ok?

tlauda commented on src/drivers/intel/cavs/timer.c:
```diff
@@ -115,36 +120,52 @@ static int platform_timer_register(struct timer *timer,
+		if (i == cpu)
+			continue;
+
+		dcache_invalidate_region(platform_generic_queue + i,
```
I think it still doesn't resolve our issue with not up to date timer->logical_irq.
Master core will register timer interrupt first, then after some time slave core will do it.
Master core still won't have invalidated data.

lyakh commented on src/drivers/intel/cavs/interrupt.c:
```diff
@@ -31,60 +32,61 @@
-	while (irq_read(ilxsd)) {
+	for (;;) {
+		unsigned int bit = ffs(status) - 1;
+		bool handled = false;
```
sorry, why? Where does this requirement come from?..

tlauda replied:
@lyakh It is the style enforced by @lgirdwood and probably comes from Linux kernel itself.

jajanusz replied:
@lyakh You won't find it in kernel guidelines, but it's enforced by maintainer and it's better to be consistent. I'm not sure but I guess it's because of c89 compatibilty.

lyakh commented on src/drivers/intel/cavs/interrupt.c:
```diff
@@ -117,49 +127,30 @@ static void irq_lvl2_level5_handler(void *data)
+		.name = irq_name_level2,
+		.irq = IRQ_NUM_EXT_LEVEL2,
+		.handler = irq_lvl2_level2_handler,
+		.global_mask = false,
```
it's been added by an earlier patch

lyakh commented on src/drivers/intel/cavs/timer.c:
```diff
@@ -115,36 +120,52 @@ static int platform_timer_register(struct timer *timer,
 
 int timer_register(struct timer *timer, void(*handler)(void *arg), void *arg)
 {
+	unsigned int cpu = cpu_get_id();
```
already there

lyakh commented on src/drivers/interrupt.c:
```diff
@@ -73,6 +75,38 @@ int interrupt_cascade_register(const struct irq_cascade_tmpl *tmpl)
+
+	for (cascade = cascade_root.list; cascade; cascade = cascade->next)
+		/* .name is non-volatile */
+		if (!rstrcmp(name, cascade->name)) {
```
yep, as discussed, we can look at that later

lyakh commented:
@tlauda to test results: the checkpatch result I'd consider a false positive, on-device I don't see any failures under the "Details" link - all tests are marked as "pass" there.

tlauda commented:
SOFCI TEST

tlauda commented:
@lyakh I've restarted device-test. Checkpatch complain maybe is a false positive, but list_for_item is used everywhere without a space, so it would be good to align with the rest of the code. Also there is some kind of problem with pausing and releasing pipelines. @zrombel will send you a FW log from the failure.

zrombel commented:
@lyakh now we are observing fails in complex stream operation tests on builds with SMP (ICL and CNL). 
Test is starting two pipelines (playback and capture) and sends TrigPause, TrigRelease sequence as shown below:

Cap Pipe TrigStart
Plb Pipe TrigStart
sleep(1s)
Cap Pipe TrigPause
Plb Pipe TrigPause
sleep(1s)
Plb Pipe TrigRelease
sleep(1s)
Plb Pipe TrigPause
Cap Pipe TrigRelease
-> FW error -17 <-

In trace log there is this message:
error: IRQ 0x90 handler argument re-used!

Full trace log below:
[dma_trace_1.log](https://github.com/thesofproject/sof/files/3421592/dma_trace_1.log)

For non-smp builds (APL) this error doesn't occur.

lyakh commented:
> @lyakh now we are observing fails in complex stream operation tests on builds with SMP (ICL and CNL).
> Test is starting two pipelines (playback and capture) and sends TrigPause, TrigRelease sequence as shown below:
> 
> Cap Pipe TrigStart
> Plb Pipe TrigStart
> sleep(1s)
> Cap Pipe TrigPause
> Plb Pipe TrigPause
> sleep(1s)
> Plb Pipe TrigRelease
> sleep(1s)
> Plb Pipe TrigPause
> Cap Pipe TrigRelease
> -> FW error -17 <-
> 
> In trace log there is this message:
> error: IRQ 0x90 handler argument re-used!
> 
> Full trace log below:
> [dma_trace_1.log](https://github.com/thesofproject/sof/files/3421592/dma_trace_1.log)
> 
> For non-smp builds (APL) this error doesn't occur.

@tlauda @zrombel This is interesting. This error means, that somebody is trying to register a DMA interrupt for the same channel for the second time. In the log I see
```
dw_dma_release(): dma 0 channel 1 release
```
and then already
```
dw_dma_interrupt_register() error: dma 0 channel 1 failed to allocate IRQ
```
and there is no
```
dw_dma_stop(): dma 0 channel 1 stop
```
as seen a few lines above, and without a .stop() you cannot allocate again, the interrupt is still busy.

tlauda commented on src/drivers/intel/cavs/interrupt.c:
```diff
@@ -173,18 +166,15 @@ void platform_interrupt_mask(uint32_t irq)
 		break;
 	case IRQ_NUM_EXT_LEVEL2:
 		irq_write(REG_IRQ_IL2MSD(core), 1 << SOF_IRQ_BIT(irq));
-		break;
```
It's error prone. Someone in the future can miss the lack of break.

tlauda commented on src/include/sof/drivers/interrupt.h:
```diff
@@ -51,28 +50,56 @@ struct irq_desc {
+ * \brief cascading interrupt controller descriptor.
+ */
+struct irq_cascade_desc {
+	const char *name;				/**< name of the
```
I think it would look better with the comment above field.

tlauda commented on src/schedule/ll_schedule.c:
```diff
@@ -545,14 +543,28 @@ static int ll_scheduler_init(void)
 
-	if (cpu_get_id() == PLATFORM_MASTER_CORE_ID) {
+	if (cpu == PLATFORM_MASTER_CORE_ID) {
+		unsigned int i;
```
Top of the function.

tlauda commented on src/schedule/ll_schedule.c:
```diff
@@ -545,14 +543,28 @@ static int ll_scheduler_init(void)
+		for (i = 0, ts = platform_generic_queue;
+		     i < PLATFORM_CORE_COUNT; i++, ts++) {
+			ts->timer.tirq = ll_shared_ctx->tirq + i;
+			dcache_writeback_region(&ts->timer.tirq,
```
It's uncached, so writeback is not needed.

plbossart commented on src/drivers/intel/cavs/interrupt.c:
```diff
@@ -196,12 +186,43 @@ void platform_interrupt_unmask(uint32_t irq)
 		irq_write(REG_IRQ_IL2MCD(core), 1 << SOF_IRQ_BIT(irq));
-		break;
-	default:
-		break;
```
same here, this is not really safe at all.

jajanusz commented on src/include/sof/common.h:
```diff
@@ -29,6 +29,8 @@
 
 #define __aligned(x) __attribute__((__aligned__(x)))
 
+#define ffs(i) __builtin_ffs(i)
```
@tlauda @lyakh  Maybe that's not the best place for this macro, just because it's __builtin. What if used compiler doesn't support it? It's more like a math function, that may be replaced with builtin if compiler allows.

lyakh replied:
@jajanusz Maybe, do you have a better idea? Do you have any specific compilers which don't implement this which we use? I'd propose to keep this until we find a better solution. FWIW it was a suggestion from Liam to put it originally in sof.h, now these macros got moved to common.h

jajanusz replied:
@lyakh It was a suggestion, so you would rethink it. It's more about what domain macro/function belong than what's supported. For now we don't support any compiler that doesn't have that feature and I don't know of any plans of doing it, so it may stay in common.h for now. Feel free to rebase so we can merge this one.

tlauda commented:
SOFCI TEST

tlauda commented:
@xiulipan CI is dead?

tlauda commented:
SOFCI TEST

lyakh commented:
@jajanusz @tlauda the CI failure
```
aplay: pcm_write:2011: write error: Input/output error
```
is real but I've just verified it's there with the current master as well, seems to be a regression. Shall I file a bug?

tlauda commented:
@lyakh It's kernel regression, right? There are no FW patches since yesterday, that could break it.

lyakh commented:
@tlauda ok, sorry, looks like I was wrong. I cannot reproduce it yet - neither with the master nor with this PR. I have tested with a wrong topology. I'll try some more to reproduce this...

lyakh commented:
@tlauda I still cannot reproduce it, I'll continue more, and it *doesn't* look like it can be related to CI network problems, that @xiulipan has recently mentioned, or can it be?

tlauda commented:
SOFCI TEST

lyakh commented:
Do I understand it correctly, that the same test has failed again, but the failure is different?..

lyakh commented:
I just looked at today's regression list: there Up2 nocodec HDMI tests are marked as "N/A" - aren't those the tests that are failing now?

jajanusz commented:
@lyakh Most likely, I'd wait for qb tests also to perform, they shouldn't have issues atm.

lyakh commented:
newest kernel + this my PR + nocodec: cannot reproduce

lyakh commented:
@jajanusz qb test?

lyakh commented:
...I also don't see any errors in traces or in dmesg, unlike the last time with https://sof-ci.01.org/sofpr/PR1053/build2601/devicetest/APL_UP2_PCM512X/check_pause_resume.sh/slogger.txt

lyakh commented:
Another observation: comparing the two tests with the same code base: https://sof-ci.01.org/sofpr/PR1053/build2604/devicetest/ and https://sof-ci.01.org/sofpr/PR1053/build2601/devicetest/ - they fail and pass on different Up2 cases - the former fails nocodec tests, the latter fails pcm512x tests.

jajanusz commented:
> @jajanusz qb test?

Internal Intel CI, you can see it in other PRs, it takes some time to run all tests there so you can wait few hours in the mid of day

lyakh commented:
@jajanusz you mean the 2 "Internal Intel CI" tests like e.g. in #1658 ? But those seem to be just code checking and building, no run-time testing?

xiulipan commented:
@lyakh The CFL multiple-pipeline-capture is timeout. could you manually check this case on some HDA codec device?

lyakh commented:
@xiulipan I don't understand this log https://sof-ci.01.org/sofpr/PR1053/build2609/devicetest/CFL_RVP_HDA/multiple-pipeline-capture.sh/multiple-pipeline-capture.sh.txt - is it a timeout? Is this what you'd like me to check?

lyakh commented:
@xiulipan I've just run
```
arecord -D hw:0,6 --dump-hw-param -c 2 -r 48000 -f s16_le /dev/null -d 5 &
arecord -D hw:0,7 --dump-hw-param -c 2 -r 16000 -f s16_le /dev/null -d 5 &
arecord -D hw:0,0 --dump-hw-param -c 2 -r 48000 -f s16_le /dev/null -d 5 &
aplay -D hw:0,3 --dump-hw-param -c 2 -r 48000 -f s16_le /dev/zero -d 5
```
on an HDA system and they all completed normally

lyakh commented:
I cannot find the actual failure in the above test at https://quickbuild.igk.intel.com/build/4062204 - it says KW found something, but there's no information what and where it found.

tlauda commented:
SOFCI TEST

tlauda commented:
@xiulipan Is this real fail or some problem with kernel?

xiulipan commented:
@tlauda @lyakh 
Not sure now. I am thinking that we have code base that can not pass multiple-pipeline-capture.sh now. Need manually check to confirm.

@ClarexZhou @Jiangxinx 
Could you help to run some manual test like multiple-pipeline-capture.sh base on this pr.

tlauda commented:
@xiulipan @ClarexZhou @Jiangxinx This is urgent. I would like to merge it as soon as possible. Could you execute manual tests?

tlauda commented:
SOFCI TEST

tlauda commented:
SOFCI TEST

lyakh commented:
@tlauda when you follow the "details" link, you get all fields as "passed"...

tlauda commented:
@lyakh @xiulipan Yes, but when I go into multiple-pipeline-capture on CFL I see some fails. I really don't know what is going on. Could someone actually confirm if it's the real issue or not?

Jiangxinx commented:
I tested BYT and APL platform, found one new issue on BYT platform:

**Failed while do playback and capture at same time:   I/O error (100%).**

By the way, apl still  can not pass multiple-pipeline-capture.sh ,and also can not pass multiple-pipeline-capture.sh on daily version. 

***Env***
Kernel: f745362c on topic/sof-dev
platform: apl-up2 & BYT

tlauda commented:
@Jiangxinx If I understand you correctly multiple-pipeline-capture on APL is no regression from this PR, but simultaneous playback and capture on BYT is.

Jiangxinx commented:
@tlauda Yes ,I have updated the simultaneous playback and capture issue information.

ClarexZhou commented:
multiple-pipeline-capture isn't a regression, also failed on daily sof on CFL. see #1674 

And test this PR on GLK i2s notebook, CML HDA notebook, CFL-S HDA RVP and ICL-HDA RVP: No regression.

Kernel(topic/sof-dev):	f745362c 
FW:	https://sof-ci-storage.sh.intel.com/pr/sof/PR1053-9/

tlauda commented:
@ClarexZhou @Jiangxinx Thanks for checking! Why the playback + capture fails on BYT, but CI reports PASS?

Jiangxinx commented:
@tlauda I think CI have not finished aplay, this new regression only can be reproduced after finished aplay.

lyakh commented:
@Jiangxinx what topology did you test with BYT? And what BYT platform? A MinnowBoard?

Jiangxinx commented:
@lyakh MinnowBoard: sof-byt-nocodec.tplg & sof-byt-rt5651.tplg

xiulipan commented:
@lyakh @tlauda @Jiangxinx @ClarexZhou 
Thanks for help testing and looking for detail in test result.

That is why I would like to manually notify here, it seem with issue #1674 we will get a timeout in real. But it seems our parser for the test result could not recognize timeout there. I have raise a bug for the test framework.

Also we find that the new added multiple capture test case may cause some test failure, not sure if this comes from code base or the PR. So I think we can check the daily build test result and then we can see if the PR have issue.

We are still debugging about the new test case to see if we are doing some invalid test.

xiulipan commented:
@lyakh @tlauda 
For the BYT issue, I checked the CI test logs. It seems we only start a capture for the test case and no other PCM at same time.

@fredoh9 
Please help to check the test case, why only start one capture. Actually we can start the playback for the same PCM.
https://sof-ci.01.org/sofpr/PR1053/build2622/devicetest/BYT_MB_RT5682/multiple-pipeline-capture.sh/multiple-pipeline-capture.sh.txt
https://sof-ci.01.org/sofpr/PR1053/build2622/devicetest/BYT_MB_NOCODEC/multiple-pipeline-capture.sh/multiple-pipeline-capture.sh.txt

tlauda commented:
@xiulipan Great, thanks for the update. Yes, we should definitely add also playback to extend our test coverage. This way we won't need such heavy manual PR testing in the future.

lyakh commented:
@Jiangxinx @xiulipan have you actually tried aplay audio on MinnowBoard with rt5651? It comes out distorted for me with the head of the current master branch (I think it's a known issue when pulseaudio is used). As for this PR - I've just tried simultaneous playback and record with my PR and with the master head and I get the same result - both work (apart from the audio quality of course). In the trace there's also a flood of
```
dw_dma_avail_data_size() size is 0! Channel enable = 0.
```
messages with both firmware versions.

lgirdwood commented:
@tlauda @lyakh @Jiangxinx @xiulipan - I'm going to merge this today as it looks like this BYT+PA issue is unrelated.

tlauda commented:
@lgirdwood Let's do this before merging any other patch, so there is no need for rebase.

lgirdwood commented:
@lyakh @tlauda done.

lyakh commented:
@lgirdwood thanks, appreciated!
